# Hiperparâmetros
- Para adiantar o processo, na proxima celula estará todo o primeiro passo da EDA

In [2]:
# Importando as metricas
import pandas as pd
from sklearn import metrics
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import confusion_matrix,recall_score,precision_score,auc
import numpy as np
from sklearn.model_selection import GridSearchCV
# Importando a base
base = pd.read_csv(r'..\data\creditcard.csv\creditcard.csv')
cols_MinMax = ['Time','Amount']
scaler = MinMaxScaler()
# Ajuste e transforme os dados das colunas selecionadas
scaled_data = scaler.fit_transform(base[cols_MinMax])
# Criando um novo DataFrame com os dados escalados
scaled_df = pd.DataFrame(scaled_data, columns=cols_MinMax)
# Substituindo as colunas na base
base[['Time', 'Amount']] = scaled_df
# Separando em X e y
X=base.drop('Class',axis=1)
y=base.Class
#TrainTestSplit
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.33, random_state=42,stratify=y)
#RUS
from imblearn.under_sampling import RandomUnderSampler
rus = RandomUnderSampler(random_state=42)
X_res, y_res = rus.fit_resample(X_train, y_train)

In [3]:
base.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean        0.548717  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std         0.274828  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%         0.313681 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%         0.490138  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%         0.806290  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max         1.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16       0.003439   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01       0.009736   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       0.000218   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02       0.000856   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02       0.003004   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01       1.000000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

### Modelos escolhidos:
- RL
- SVC
- KNN
- RF
#### RL

In [5]:
from sklearn.linear_model import LogisticRegression
clf_RL = LogisticRegression(random_state=42).fit(X_res, y_res)
y_pred_RL = clf_RL.predict(X_test)
y_pred_proba_RL = clf_RL.predict_proba(X_test)[:,1]

In [6]:
confusion_matrix(y_test, y_pred_RL)

array([[89868,  3957],
       [   15,   147]], dtype=int64)

In [7]:
import warnings
warnings.filterwarnings('ignore')
# Definindo os valores de C (regularização) para explorar
#C_values = np.logspace(-2, 3, num=6)
# Definindo os parâmetros para o GridSearch
parameters = {
    'C': [0.001,0.01, 0.1, 1, 10, 100, 1000], 
    'solver': ['liblinear', 'saga', 'newton-cg', 'lbfgs','sag'],  
}

# Criando o classificador de Regressão Logística
RL_GS = LogisticRegression(random_state=42)
clf_GS_LR = GridSearchCV(RL_GS, 
                             parameters, 
                             scoring='recall').fit(X_res, y_res)

In [8]:
# Verificando quais os melhores parametros para o SVM
clf_GS_LR.best_params_

{'C': 0.001, 'solver': 'liblinear'}

In [9]:
#Fazendo a predição com os melhores parametros
y_pred_RL_GS = clf_GS_LR.predict(X_test)

In [10]:
#Usando a confusion Matrix
confusion_matrix(y_test, y_pred_RL_GS)

array([[81700, 12125],
       [   10,   152]], dtype=int64)

- Neste primeiro caso, pioramos o os casos que de falso positivos, mas melhoramos os verdadeiros positivos.
#### SVC

In [12]:
from sklearn.svm import SVC
clf_SVC = SVC(random_state=42,probability=True).fit(X_res, y_res)
y_pred_SVC = clf_SVC.predict(X_test)
y_pred_proba_SVC = clf_SVC.predict_proba(X_test)[:,1]

In [13]:
confusion_matrix(y_test, y_pred_SVC)

array([[92433,  1392],
       [   24,   138]], dtype=int64)

In [14]:
C_values = np.logspace(-2, 2, num=5)
C_values

array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02])

In [15]:
from sklearn.model_selection import GridSearchCV
# Para variar o valor de C de 0.01 até 1000
C_values = np.logspace(-2, 3, num=6)
#Escolhendo os parametros 
parameters = {'kernel':('linear', 'rbf','poly','sigmoid'), 
              'C':C_values,
             'class_weight': [None, 'balanced']}
#Criando um classificado e fazendo o fit
svc_GS = SVC(random_state=42,probability=True)
clf_GS_SVM = GridSearchCV(svc_GS, 
                          parameters, 
                          scoring='recall').fit(X_res,y_res)

In [16]:
# Verificando quais os melhores parametros para o SVM
clf_GS_SVM.best_params_

{'C': 1000.0, 'class_weight': None, 'kernel': 'linear'}

In [17]:
#Fazendo a predição com os melhores parametros
y_pred_SVM_GS = clf_GS_SVM.predict(X_test)

In [18]:
#Usando a confusion Matrix
confusion_matrix(y_test, y_pred_SVM_GS)

array([[89098,  4727],
       [   14,   148]], dtype=int64)

- Comparado ao RL, melhoramos o os casos que de falso positivos, mas pioramos os verdadeiros positivos
#### KNN

In [20]:
from sklearn.neighbors import KNeighborsClassifier
clf_KNN = KNeighborsClassifier().fit(X_res, y_res)
y_pred_KNN = clf_KNN.predict(X_test)
y_pred_proba_KNN = clf_KNN.predict_proba(X_test)[:,1]

In [21]:
confusion_matrix(y_test, y_pred_KNN)

array([[91929,  1896],
       [   19,   143]], dtype=int64)

In [22]:
#Escolhendo os parametros 

parameters = {'n_neighbors':[3, 5, 7, 9], 
              'weights':("uniform", "distance"),
             'algorithm': ("auto", "ball_tree", "kd_tree", "brute")}
#Criando um classificado e fazendo o fit
KNN_GS = KNeighborsClassifier()
clf_GS_KNN = GridSearchCV(KNN_GS, 
                          parameters, 
                          scoring='recall').fit(X_res,y_res)

In [23]:
# Verificando quais os melhores parametros para o SVM
clf_GS_KNN.best_params_

{'algorithm': 'auto', 'n_neighbors': 3, 'weights': 'uniform'}

In [24]:
#Fazendo a predição com os melhores parametros
y_pred_KNN_GS = clf_GS_KNN.predict(X_test)

In [25]:
#Usando a confusion Matrix
confusion_matrix(y_test, y_pred_KNN_GS)

array([[91136,  2689],
       [   15,   147]], dtype=int64)

- Em comparação com o SVC, melhoramos os verdadeiros positivos, porém em comparação ao KNN pioramos os Falso positivo.
#### RF

In [27]:
from sklearn.ensemble import RandomForestClassifier
clf_RF =RandomForestClassifier(max_depth=2, random_state=42).fit(X_res, y_res)
y_pred_RF = clf_RF.predict(X_test)
y_pred_proba_RF = clf_RF.predict_proba(X_test)[:,1]

In [28]:
confusion_matrix(y_test, y_pred_RF)

array([[93221,   604],
       [   26,   136]], dtype=int64)

In [29]:
#Escolhendo os parametros 

parameters = {'n_estimators':[50, 100, 200, 400], 
              'criterion':("gini", "entropy", "log_loss"),
             'max_depth': [1,3,5,7,9,12]}
#Criando um classificado e fazendo o fit
RF_GS = RandomForestClassifier(random_state=42)
clf_GS_RF = GridSearchCV(RF_GS, 
                          parameters, 
                          scoring='recall').fit(X_res,y_res)

In [30]:
# Verificando quais os melhores parametros para o SVM
clf_GS_RF.best_params_

{'criterion': 'gini', 'max_depth': 9, 'n_estimators': 200}

In [31]:
#Fazendo a predição com os melhores parametros
y_pred_RF_GS = clf_GS_RF.predict(X_test)

In [32]:
#Usando a confusion Matrix
confusion_matrix(y_test, y_pred_RF_GS)

array([[91531,  2294],
       [   19,   143]], dtype=int64)

In [33]:
#Usando a confusion Matrix
confusion_matrix(y_test, y_pred_RF_GS)

array([[91531,  2294],
       [   19,   143]], dtype=int64)

- Em relação ao KNN, uma melhora nos falso positivos, porém uma piora nos verdadeiros positivos.
### Decisão:
- Por se tratar de um problema de identificação de fraude, optei por utilizar o modelo de Regressão Linear, por ter sido o modelo que mais teve verdadeiros positivos (152).
- Um segundo e cenário seria, "a empresa irá ligar para todos os clientes que o algoritimo prever como fraude", logo nesse caso, se o objetivo principal é reduzir o número de falsas positivas para evitar ligações desnecessárias e custos associados, o RandomForest é a melhor escolha.

### Tentando adicionar mais hiperparâmetros em RL

In [35]:
from sklearn.linear_model import LogisticRegression
clf_RL = LogisticRegression(random_state=42).fit(X_res, y_res)
y_pred_RL = clf_RL.predict(X_test)
y_pred_proba_RL = clf_RL.predict_proba(X_test)[:,1]

In [36]:
warnings.filterwarnings('ignore')
# Definindo os valores de C (regularização) para explorar
#C_values = np.logspace(-2, 3, num=6)
# Definindo os parâmetros para o GridSearch
parameters = {
    'C': [0.001,0.01, 0.1, 1, 10, 100, 1000],
    'penalty': ['l1', 'l2','elasticnet',None], 
    'solver': ['liblinear', 'saga', 'newton-cg', 'lbfgs','sag','newton-cholesky'], 
    'max_iter':[100,150,200, 1000]
}

# Criando o classificador de Regressão Logística
RL_GS = LogisticRegression(random_state=42)
clf_GS_LR = GridSearchCV(RL_GS, 
                             parameters, 
                             scoring='recall').fit(X_res, y_res)

In [37]:
# Verificando quais os melhores parametros para o SVM
clf_GS_LR.best_params_

{'C': 0.01, 'max_iter': 100, 'penalty': 'l1', 'solver': 'liblinear'}

In [38]:
#Fazendo a predição com os melhores parametros
y_pred_RL_GS = clf_GS_LR.predict(X_test)

In [39]:
#Usando a confusion Matrix
confusion_matrix(y_test, y_pred_RL_GS)

array([[81095, 12730],
       [   10,   152]], dtype=int64)

In [40]:
recall_score(y_test,y_pred_RL_GS)

0.9382716049382716

- Adicionando mais hiperparâmetros não melhorou nosso modelo, logo manteremos o primeiro modelo.
- Para tentar melhorar o modelo, utilizarei outras formas de undersampling
### NearMiss

In [42]:
from imblearn.under_sampling import NearMiss
nm = NearMiss()
X_resNM, y_resNM = nm.fit_resample(X_train, y_train)

In [43]:
warnings.filterwarnings('ignore')
# Definindo os valores de C (regularização) para explorar
#C_values = np.logspace(-2, 3, num=6)
# Definindo os parâmetros para o GridSearch
parameters = {
    'C': [0.001,0.01, 0.1, 1, 10, 100, 1000],
    'penalty': ['l1', 'l2','elasticnet',None], 
    'solver': ['liblinear', 'saga', 'newton-cg', 'lbfgs','sag','newton-cholesky'], 
    'max_iter':[100,150,200, 1000]
}

# Criando o classificador de Regressão Logística
RL_GS = LogisticRegression(random_state=42)
clf_GS_LR = GridSearchCV(RL_GS, 
                             parameters, 
                             scoring='recall').fit(X_resNM, y_resNM)

In [44]:
# Verificando quais os melhores parametros para o SVM
clf_GS_LR.best_params_

{'C': 1000, 'max_iter': 100, 'penalty': 'l2', 'solver': 'newton-cholesky'}

In [45]:
#Fazendo a predição com os melhores parametros
y_pred_RL_GS = clf_GS_LR.predict(X_test)

In [46]:
#Usando a confusion Matrix
confusion_matrix(y_test, y_pred_RL_GS)

array([[49285, 44540],
       [   12,   150]], dtype=int64)

- Aumentaram bastante os resultados de falso positivos sem ganhos em outro quadrante.
### Cluster Centroids

In [48]:
from imblearn.under_sampling import ClusterCentroids
cc = ClusterCentroids(random_state=42)
X_resCC, y_resCC = cc.fit_resample(X_train, y_train)

In [49]:
import warnings
warnings.filterwarnings('ignore')
# Definindo os valores de C (regularização) para explorar
#C_values = np.logspace(-2, 3, num=6)
# Definindo os parâmetros para o GridSearch
parameters = {
    'C': [0.001,0.01, 0.1, 1, 10, 100, 1000], 
    'solver': ['liblinear', 'saga', 'newton-cg', 'lbfgs','sag'],  
}

# Criando o classificador de Regressão Logística
RL_GS = LogisticRegression(random_state=42)
clf_GS_LR = GridSearchCV(RL_GS, 
                             parameters, 
                             scoring='recall').fit(X_resCC, y_resCC)

In [50]:
# Verificando quais os melhores parametros para o RL
clf_GS_LR.best_params_

{'C': 0.001, 'solver': 'liblinear'}

In [51]:
#Fazendo a predição com os melhores parametros
y_pred_RL_GS = clf_GS_LR.predict(X_test)

In [52]:
#Usando a confusion Matrix
confusion_matrix(y_test, y_pred_RL_GS)

array([[74026, 19799],
       [   13,   149]], dtype=int64)

- Em Relação ao NearMiss, uma melhora consideravel com os falsos positivos, porém ainda pior que o modelo de RUS.
#### Tentanto melhorar o GridSearch da RL inicial adicionando a Precisão

In [54]:
import warnings
warnings.filterwarnings('ignore')
# Definindo os valores de C (regularização) para explorar
#C_values = np.logspace(-2, 3, num=6)
# Definindo os parâmetros para o GridSearch
parameters = {
    'C': [0.001,0.01, 0.1, 1, 10, 100, 1000], 
    'solver': ['liblinear', 'saga', 'newton-cg', 'lbfgs','sag'],  
}

# Criando o classificador de Regressão Logística
RL_GS = LogisticRegression(random_state=42)
clf_GS_LR = GridSearchCV(RL_GS, 
                             parameters, 
                             scoring=['recall','precision'],refit='recall').fit(X_res, y_res)

In [55]:
# Verificando quais os melhores parametros para o RL
clf_GS_LR.best_params_

{'C': 0.001, 'solver': 'liblinear'}

In [56]:
#Fazendo a predição com os melhores parametros
y_pred_RL_GS = clf_GS_LR.predict(X_test)

In [57]:
#Usando a confusion Matrix
confusion_matrix(y_test, y_pred_RL_GS)

array([[81700, 12125],
       [   10,   152]], dtype=int64)

In [58]:
#Metrica final do nosso modelo
print(recall_score(y_test,y_pred_RL_GS))
print(precision_score(y_test,y_pred_RL_GS))

0.9382716049382716
0.012380874806548831


# Supondo e demonstrando preço de fraudes

In [60]:
cm = confusion_matrix(y_test, y_pred_RL_GS)
# Supondo que cada fraude custa 1000
preco_fraude = 1000
# E que o custo da análise custa 0.25
preço_analise = 0.25
# O custo de rodar o modelo custa:
custo_do_modelo = cm[1][0] *preco_fraude + cm[0][1]*preço_analise
print(f'Esse modelo custa R$ {custo_do_modelo} para a empresa')

Esse modelo custa R$ 13031.25 para a empresa
